File: service-b/src/metrics.ts
/**
 * Prometheus Metrics for Service B (Trade Execution Service)
 *
 * Exposes application metrics for monitoring trade processing performance
 */

import { Registry, Counter, Histogram, collectDefaultMetrics } from 'prom-client';

// Create a custom registry
export const register = new Registry();

// Collect default Node.js metrics (memory, CPU, event loop, etc.)
collectDefaultMetrics({ register });

/**
 * Counter: Total trades processed
 *
 * Labels:
 * - action: BUY or SELL
 * - asset_id: Trading asset identifier
 * - status: approved, rejected, error
 */
export const tradesProcessedTotal = new Counter({
  name: 'trades_processed_total',
  help: 'Total number of trades processed by Service B',
  labelNames: ['action', 'asset_id', 'status'],
  registers: [register],
});

/**
 * Histogram: Trade processing duration
 *
 * Tracks time spent processing each trade (including risk check, Redis ops, etc.)
 *
 * Buckets: 0.01s, 0.05s, 0.1s, 0.5s, 1s, 2s, 5s
 */
export const processingDurationSeconds = new Histogram({
  name: 'processing_duration_seconds',
  help: 'Time spent processing a trade signal in seconds',
  labelNames: ['action', 'asset_id', 'status'],
  buckets: [0.01, 0.05, 0.1, 0.5, 1.0, 2.0, 5.0],
  registers: [register],
});

/**
 * Counter: Risk check results
 *
 * Labels:
 * - result: allowed, rejected, error
 */
export const riskChecksTotal = new Counter({
  name: 'risk_checks_total',
  help: 'Total number of risk checks performed',
  labelNames: ['result'],
  registers: [register],
});

/**
 * Counter: Redis cache hits/misses
 *
 * Labels:
 * - operation: price_cache, trade_history
 * - result: hit, miss
 */
export const redisCacheTotal = new Counter({
  name: 'redis_cache_total',
  help: 'Total number of Redis cache operations',
  labelNames: ['operation', 'result'],
  registers: [register],
});

File: service-c/internal/metrics/metrics.go
// Package metrics provides Prometheus instrumentation for Service C (Risk Checker)
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	// RiskChecksTotal tracks the total number of risk checks performed
	// Labels: result (allowed, rejected, error), asset_id
	RiskChecksTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "risk_checks_total",
			Help: "Total number of risk checks performed by Service C",
		},
		[]string{"result", "asset_id"},
	)

	// RiskCheckDurationSeconds tracks the duration of risk check operations
	// Labels: result (allowed, rejected, error), asset_id
	// Buckets: 0.001s (1ms), 0.005s (5ms), 0.01s (10ms), 0.025s (25ms), 0.05s (50ms), 0.1s (100ms)
	RiskCheckDurationSeconds = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "risk_check_duration_seconds",
			Help:    "Time spent processing a risk check in seconds",
			Buckets: []float64{0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5},
		},
		[]string{"result", "asset_id"},
	)

	// RiskCheckVolumeHistogram tracks the distribution of trade volumes checked
	// Buckets designed around the 90 threshold (10, 25, 50, 75, 90, 100, 150, 200)
	RiskCheckVolumeHistogram = promauto.NewHistogram(
		prometheus.HistogramOpts{
			Name:    "risk_check_volume_mwh",
			Help:    "Distribution of trade volumes checked (in MWh)",
			Buckets: []float64{10, 25, 50, 75, 90, 100, 150, 200, 300},
		},
	)
)
